src/odometry_publisher.cpp

```diff
@@ -0,0 +1,215 @@
+    geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(th);
+
+    // Transform for odom/base_link, x and y inverted because different axis
+    // used in our calculations of odom.
+    geometry_msgs::TransformStamped odom_trans;
+    odom_trans.header.stamp = curTime;
+    odom_trans.header.frame_id = "odom";
+    odom_trans.child_frame_id = "base_link";
+    odom_trans.transform.translation.x = y/100.0;
+    odom_trans.transform.translation.y = -x/100.0;
+    odom_trans.transform.translation.z = 0.0;
+    odom_trans.transform.rotation = odom_quat;
+
+    // send the transform
+    odom_broadcaster.sendTransform(odom_trans);
+
+    // The actual odom message
+    nav_msgs::Odometry odom;
+    odom.header.stamp = curTime;
+    odom.header.frame_id = "odom";
+    odom.pose.pose.position.x = y/100.0;
+    odom.pose.pose.position.y = -x/100.0;
+    odom.pose.pose.position.z = 0.0;
+    odom.pose.pose.orientation = odom_quat;
+    odom.child_frame_id = "base_link";
+    odom.twist.twist.linear.x = (dy/100.0) / dt;
+    odom.twist.twist.linear.y = (-dx/100.0) / dt;
+    odom.twist.twist.angular.z = dTheta / dt;
+
+    // publish the message
+    odom_pub.publish(odom);
+}
+
+/**
+ * Returns the absolute value of a double.
+ * @param x - the input double
+ * @return - the absolute value of the input
+ */
+double OdometryPublisher::myAbs(double x) {
+    if(x < 0) {
+        return -1*x;
+    }
+    return x;
+}
+
+/**
+ * Update internal odom values;
+ */
+void OdometryPublisher::updateOdometry() {
+    // Get new position of the motors
+    double newLeftPos = base->getLeftPos();
+    double newRightPos = base->getRightPos();
+    double dLeftPos = newLeftPos - leftPos;
+    double dRightPos = newRightPos - rightPos;
+    
+    bool turnInPlace = false;
+
+    std::cout << "dRightPos: " << dRightPos << std::endl;
+
+    // left overflow
+    if (myAbs(leftPos - newLeftPos) > overflowThreshold) {
+        if(leftPos > 0) {
+            dLeftPos = dLeftPos + 2*overflowPoint;
+        } else {
+            dLeftPos = dLeftPos - 2*overflowPoint;
+        }
+    }
+
+    // right overflow
+    if (myAbs(rightPos - newRightPos) > overflowThreshold) {
+        if(rightPos > 0) {
+            dRightPos = dRightPos + 2*overflowPoint;
+        } else {
+            dRightPos = dRightPos - 2*overflowPoint;
+        }
+    }
+
+    std::cout << "dRightPos: " << dRightPos << std::endl;
+    std::cout << "getWheelDiameter: " << base->getWheelDiameter();
+
+    // Calculate distances
+    leftPos = newLeftPos;
+    rightPos = newRightPos;
+    double dLeftDistance = dLeftPos * base->getWheelDiameter() / 2.0d;
+    double dRightDistance = dRightPos * base->getWheelDiameter() / 2.0d;
+    std::cout << "dRightDistance: " << dRightDistance << std::endl;
+    leftDistance += dLeftDistance;
+    rightDistance += dRightDistance;
+
+    // Calculate radius
+    double radius;
+    if(myAbs(dLeftDistance - dRightDistance) < 0.00001d) {
```

Comment:
(A) the d in 0.000....001d and other cases is redundant, default floating point values in cpp are double.
(b) maybe define the number somewhere.